File: hamersapp/src/main/java/nl/ecci/hamers/events/SingleEventActivity.java
package nl.ecci.hamers.events;

import android.content.ContentUris;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.provider.CalendarContract;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.ActionBar;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

import com.android.volley.VolleyError;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.Date;

import nl.ecci.hamers.MainActivity;
import nl.ecci.hamers.R;
import nl.ecci.hamers.helpers.HamersActivity;
import nl.ecci.hamers.loader.Loader;
import nl.ecci.hamers.loader.VolleyCallback;
import nl.ecci.hamers.users.User;

import static nl.ecci.hamers.helpers.Utils.getEvent;
import static nl.ecci.hamers.helpers.Utils.getOwnUser;
import static nl.ecci.hamers.helpers.Utils.getUser;

public class SingleEventActivity extends HamersActivity {

    private Event event;
    private LayoutInflater inflater;
    private LinearLayout presentView;
    private LinearLayout absentView;
    private ViewGroup eventLayout;
    private ViewGroup presentLayout;
    private ViewGroup absentLayout;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.single_event);

        inflater = getLayoutInflater();

        initToolbar();

        TextView titleTV = (TextView) findViewById(R.id.event_title);
        View dateRow = findViewById(R.id.date_row);
        View locationRow = findViewById(R.id.location_row);
        View descriptionRow = findViewById(R.id.description_row);
        LinearLayout button_layout = (LinearLayout) findViewById(R.id.button_layout);
        presentView = (LinearLayout) findViewById(R.id.present_insert_point);
        absentView = (LinearLayout) findViewById(R.id.absent_insert_point);
        LinearLayout buttonLayout = (LinearLayout) findViewById(R.id.buttonLayout);

        eventLayout = (ViewGroup) findViewById(R.id.single_event_layout);
        presentLayout = (ViewGroup) findViewById(R.id.present_layout);
        absentLayout = (ViewGroup) findViewById(R.id.absent_layout);

        event = getEvent(MainActivity.prefs, getIntent().getIntExtra(Event.EVENT, 1));

        initSignups();

        titleTV.setText(event.getTitle());
        fillDetailRow(descriptionRow, event.getDescription());

        // Current date
        Date today = new Date();
        // Event deadline
        Date deadline = event.getDeadline();

        if (deadline != null) {
            if (today.after(deadline)) {
                button_layout.setVisibility(View.GONE);
                ScrollView scrollView = (ScrollView) findViewById(R.id.single_event_scrollview);
                scrollView.removeView(buttonLayout);
            }
        }

        if (dateRow != null) {
            fillImageRow(dateRow, "Datum", MainActivity.appDF.format(event.getDate()), ContextCompat.getDrawable(this, R.drawable.ic_event));
            dateRow.setClickable(true);
            dateRow.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    long startMillis = event.getDate().getTime();
                    Uri.Builder builder = CalendarContract.CONTENT_URI.buildUpon();
                    builder.appendPath("time");
                    ContentUris.appendId(builder, startMillis);
                    Intent intent = new Intent(Intent.ACTION_VIEW).setData(builder.build());
                    startActivity(intent);
                }
            });
        }

        if (locationRow != null) {
            if (event.getLocation() != null) {
                fillImageRow(locationRow, "Locatie", event.getLocation(), ContextCompat.getDrawable(this, R.drawable.location));

                locationRow.setClickable(true);
                locationRow.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        // Create a Uri from an intent string. Use the result to create an Intent.
                        Uri uri = Uri.parse("geo:0,0?q=" + event.getLocation());
                        // Create an Intent from uri. Set the action to ACTION_VIEW
                        Intent intent = new Intent(Intent.ACTION_VIEW, uri);
                        intent.setPackage("com.google.android.apps.maps");
                        if (intent.resolveActivity(getPackageManager()) != null) {
                            startActivity(intent);
                        }
                    }
                });
            } else {
                locationRow.setVisibility(View.GONE);
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.edit_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.edit_item:
                Intent intent = new Intent(this, NewEventActivity.class);
                if (event != null) {
                    intent.putExtra(Event.EVENT, event.getID());
                }
                startActivity(intent);
                return true;
        }
        return super.onOptionsItemSelected(item);
    }

    public void setPresent(View view) {
        postSignup(event.getID(), "true");
    }

    public void setAbsent(View view) {
        postSignup(event.getID(), "false");
    }

    private void postSignup(int eventID, String status) {
        JSONObject body = new JSONObject();
        try {
            body.put("event_id", eventID);
            body.put("status", status);
        } catch (JSONException ignored) {
        }

        Loader.postOrPatchData(new VolleyCallback() {
            @Override
            public void onSuccess(JSONArray response) {

            }

            @Override
            public void onError(VolleyError error) {

            }
        }, this, MainActivity.prefs, Loader.SIGNUPURL, -1, body);
        this.finish();
    }

    private void initToolbar() {
        final Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        final ActionBar actionBar = getSupportActionBar();

        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setHomeButtonEnabled(true);
        }
    }

    private void initSignups() {
        Button presentButton = (Button) findViewById(R.id.present_button);
        Button absentButton = (Button) findViewById(R.id.absent_button);
        ArrayList<Event.Signup> signups = event.getSignups();
        ArrayList<String> present = new ArrayList<>();
        ArrayList<String> absent = new ArrayList<>();
        for (int i = 0; i < signups.size(); i++) {
            Event.Signup signup = signups.get(i);
            if (signup.isAttending()) {
                present.add(getUser(MainActivity.prefs, signup.getUserID()).getName());
            } else {
                absent.add(getUser(MainActivity.prefs, signup.getUserID()).getName());
            }
        }

        if (present.size() != 0) {
            for (String name : present) {
                presentView.addView(newSingleRow(name, presentView), new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
            }
        } else {
            eventLayout.removeView(presentLayout);
        }

        if (absent.size() != 0) {
            for (String name : absent) {
                absentView.addView(newSingleRow(name, absentView), new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
            }
        } else {
            eventLayout.removeView(absentLayout);
        }

        User ownUser = getOwnUser(MainActivity.prefs);
        if (present.contains(ownUser.getName()) && presentButton != null) {
            presentButton.setVisibility(View.GONE);
        } else if (absent.contains(ownUser.getName()) && absentButton != null) {
            absentButton.setVisibility(View.GONE);
        }
    }

    private View newSingleRow(final String title, ViewGroup viewGroup) {
        View view = inflater.inflate(R.layout.row_singleview, viewGroup, false);

        TextView titleView = (TextView) view.findViewById(R.id.row_title);
        titleView.setText(title);

        return view;
    }

    private void fillDetailRow(View view, final String description) {
        TextView titleView = (TextView) view.findViewById(R.id.row_title);
        titleView.setText(R.string.description);

        TextView descriptionView = (TextView) view.findViewById(R.id.row_description);
        descriptionView.setText(description);
    }

    private void fillImageRow(View view, final String title, final String description, final Drawable image) {
        TextView titleView = (TextView) view.findViewById(R.id.row_title);
        titleView.setText(title);

        TextView descriptionView = (TextView) view.findViewById(R.id.row_description);
        descriptionView.setText(description);

        ImageView iconView = (ImageView) view.findViewById(R.id.icon);
        iconView.setImageDrawable(image);
    }
}

File: hamersapp/src/main/java/nl/ecci/hamers/events/SingleEventActivity.kt
package nl.ecci.hamers.events

import android.content.ContentUris
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.preference.PreferenceManager
import android.provider.CalendarContract
import android.support.v7.app.AlertDialog
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.Toast
import com.android.volley.VolleyError
import kotlinx.android.synthetic.main.activity_detail_item.*
import kotlinx.android.synthetic.main.stub_detail_event.*
import nl.ecci.hamers.BuildConfig
import nl.ecci.hamers.MainActivity
import nl.ecci.hamers.R
import nl.ecci.hamers.helpers.DataUtils
import nl.ecci.hamers.helpers.HamersActivity
import nl.ecci.hamers.helpers.Utils
import nl.ecci.hamers.loader.Loader
import nl.ecci.hamers.loader.PostCallback
import nl.ecci.hamers.users.User
import org.json.JSONException
import org.json.JSONObject
import java.util.*

class SingleEventActivity : HamersActivity() {

    private var event: Event? = null
    private var ownUser: User? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_detail_item)

        initToolbar()

        stub_detail_item.layoutResource = R.layout.stub_detail_event
        stub_detail_item.inflate()

        ownUser = DataUtils.getOwnUser(this)

        // Intent by clicking an event in EventFragment or by clicking a link elsewhere
        val appLinkData = intent.data
        val eventID = intent.getIntExtra(Event.EVENT, Utils.notFound)
        if (eventID != Utils.notFound) {
            event = DataUtils.getEvent(this, eventID)
        } else if (appLinkData != null) {
            event = DataUtils.getEvent(this, Utils.getIdFromUri(appLinkData))
        }

        var eventTitle = event?.title
        if (BuildConfig.DEBUG) {
            eventTitle += " (" + event?.id + ")"
        }

        event_title.text = eventTitle
        fillDetailRow(description_row, getString(R.string.description), event?.description)

        if (Date().after(event?.deadline)) {
            button_layout.visibility = View.GONE
            single_event_scrollview.removeView(button_layout)
        }

        fillImageRow(date_row, "Datum", MainActivity.appDF.format(event!!.date), R.drawable.ic_event)
        date_row.isClickable = true
        date_row.setOnClickListener {
            val startMillis = event!!.date.time
            val builder = CalendarContract.CONTENT_URI.buildUpon()
            builder.appendPath("time")
            ContentUris.appendId(builder, startMillis)
            val intent = Intent(Intent.ACTION_VIEW).setData(builder.build())
            startActivity(intent)
        }

        if (event!!.location.isNotEmpty()) {
            fillImageRow(location_row, "Locatie", event!!.location, R.drawable.location)

            location_row.isClickable = true
            location_row.setOnClickListener {
                // Create a Uri from an intent string. Use the result to create an Intent.
                val uri = Uri.parse("geo:0,0?q=" + event!!.location)
                // Create an Intent from uri. Set the action to ACTION_VIEW
                val intent = Intent(Intent.ACTION_VIEW, uri)
                intent.`package` = "com.google.android.apps.maps"
                if (intent.resolveActivity(packageManager) != null) {
                    startActivity(intent)
                }
            }
        } else {
            location_row.visibility = View.GONE
        }

        present_button.setOnClickListener { postSignUp(true, null) }
        if (event!!.attendance) {
            absent_button.setOnClickListener { askForReason() }
        } else {
            absent_button.setOnClickListener { postSignUp(false, null) }
        }

        initSignUps()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.edit_menu, menu)
        if (ownUser?.id != event!!.userID) {
            menu.removeItem(R.id.edit_item)
        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }
            R.id.edit_item -> {
                Intent(this, NewEventActivity::class.java).putExtra(Event.EVENT, event?.id)
                startActivity(intent)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun postSignUp(status: Boolean?, reason: String?) {
        val body = JSONObject()
        try {
            body.put("event_id", event!!.id)
            body.put("status", status!!.toString())

            // Attendance is mandatory, so ask for the reason for absence!
            if (!status && event!!.attendance && reason != null) {
                body.put("reason", reason)
            }
        } catch (ignored: JSONException) {
        }

        Loader.postOrPatchData(this, Loader.SIGNUPURL, body, -1, object : PostCallback {
            override fun onSuccess(response: JSONObject) {
                finish()
            }

            override fun onError(error: VolleyError) {

            }
        })
    }

    private fun initSignUps() {
        val signUps = event?.signUps
        val present = ArrayList<String>()
        val absent = ArrayList<String>()
        signUps?.indices
                ?.map { signUps[it] }
                ?.forEach {
                    if (it.isAttending) {
                        present.add(DataUtils.getUser(this, it.userID).name)
                    } else {
                        absent.add(DataUtils.getUser(this, it.userID).name)
                    }
                }

        if (present.size > 0) {
            for (name in present) {
                present_insert_point.addView(newSingleRow(name, present_insert_point), ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT))
            }
            event_present_count.text = present.size.toString()
        } else {
            single_event_layout.removeView(present_layout)
        }

        if (absent.size > 0) {
            for (name in absent) {
                absent_insert_point.addView(newSingleRow(name, absent_insert_point), ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT))
            }
            event_absent_count.text = absent.size.toString()
        } else {
            single_event_layout.removeView(absent_layout)
        }

        if (present.contains(ownUser?.name)) {
            present_button.visibility = View.GONE
        } else if (absent.contains(ownUser?.name)) {
            absent_button.visibility = View.GONE
        }
    }

    private fun askForReason() {
        val alert = AlertDialog.Builder(this)
        alert.setTitle(R.string.attendance_reason_title)

        val input = EditText(this)
        input.setSingleLine()
        input.setHint(R.string.attendance_reason_message)

        val container = FrameLayout(this)
        container.addView(input)

        val params = FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
        params.marginStart = resources.getDimensionPixelSize(R.dimen.dialog_margin)
        params.marginEnd = resources.getDimensionPixelSize(R.dimen.dialog_margin)
        input.layoutParams = params

        alert.setView(container)

        alert.setPositiveButton(android.R.string.yes) { _, _ ->
            val reason = input.text.toString()
            if (reason.length > 5) {
                postSignUp(false, reason)
            } else {
                Toast.makeText(this@SingleEventActivity, R.string.attendance_reason_size, Toast.LENGTH_SHORT).show()
            }
        }
        alert.setNegativeButton(android.R.string.no) { _, _ ->
            // Do nothing.
        }

        alert.show()
    }

}
